File: Objective-C/advanced/APIDemo/APIDemo/Snippets/RewardedInterstitialAdSnippets.m
//
//  Copyright (C) 2025 Google, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

#import <Foundation/Foundation.h>
#import <GoogleMobileAds/GoogleMobileAds.h>

// Replace this ad unit ID with your own ad unit ID.
static NSString *const adUnitID = @"ca-app-pub-3940256099942544/6978759866";
static NSString *const adManagerAdUnitID = @"/21775744923/example/rewarded-interstitial";

@interface RewardedInterstitialAdSnippets : UIViewController <GADFullScreenContentDelegate>

@property(strong, nonatomic) GADRewardedInterstitialAd *rewardedInterstitialAd;

@end

@implementation RewardedInterstitialAdSnippets

// [START load_ad]
- (void)loadRewardedInterstitialAd {
  [GADRewardedInterstitialAd loadWithAdUnitID:adUnitID
                                      request:[GADRequest request]
                            completionHandler:^(GADRewardedInterstitialAd *ad, NSError *error) {
                              if (error) {
                                NSLog(@"Failed to load rewarded interstitial ad with error: %@",
                                      error.localizedDescription);
                                return;
                              }
                              self.rewardedInterstitialAd = ad;
                              // [START set_delegate]
                              self.rewardedInterstitialAd.fullScreenContentDelegate = self;
                              // [END set_delegate]
                            }];
}
// [END load_ad]

// [START load_ad_gam]
- (void)loadAdManagerRewardedInterstitialAd {
  [GADRewardedInterstitialAd loadWithAdUnitID:adUnitID
                                      request:[GAMRequest request]
                            completionHandler:^(GADRewardedInterstitialAd *ad, NSError *error) {
                              if (error) {
                                NSLog(@"Failed to load rewarded interstitial ad with error: %@",
                                      error.localizedDescription);
                                return;
                              }
                              self.rewardedInterstitialAd = ad;
                              self.rewardedInterstitialAd.fullScreenContentDelegate = self;
                            }];
}
// [END load_ad_gam]

// [START validate_server_side_verification]
- (void)validateServerSideVerification {
  // Replace this ad unit ID with your own ad unit ID.
  [GADRewardedInterstitialAd loadWithAdUnitID:adUnitID
                                      request:[GADRequest request]
                            completionHandler:^(GADRewardedInterstitialAd *ad, NSError *error) {
                              if (error) {
                                NSLog(@"Rewarded interstitial ad failed to load with error: %@",
                                      error.localizedDescription);
                                return;
                              }
                              self.rewardedInterstitialAd = ad;
                              GADServerSideVerificationOptions *options =
                                  [[GADServerSideVerificationOptions alloc] init];
                              options.customRewardString = @"SAMPLE_CUSTOM_DATA_STRING";
                              ad.serverSideVerificationOptions = options;
                            }];
}
// [END validate_server_side_verification]

// [START validate_gam_server_side_verification]
- (void)validateAdManagerServerSideVerification {
  // Replace this ad unit ID with your own ad unit ID.
  [GADRewardedInterstitialAd loadWithAdUnitID:adUnitID
                                      request:[GAMRequest request]
                            completionHandler:^(GADRewardedInterstitialAd *ad, NSError *error) {
                              if (error) {
                                NSLog(@"Rewarded interstitial ad failed to load with error: %@",
                                      error.localizedDescription);
                                return;
                              }
                              self.rewardedInterstitialAd = ad;
                              GADServerSideVerificationOptions *options =
                                  [[GADServerSideVerificationOptions alloc] init];
                              options.customRewardString = @"SAMPLE_CUSTOM_DATA_STRING";
                              ad.serverSideVerificationOptions = options;
                            }];
}
// [END validate_gam_server_side_verification]

#pragma mark GADFullScreeContentDelegate implementation

// [START ad_events]
- (void)adDidRecordImpression:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"%s called", __PRETTY_FUNCTION__);
}

- (void)adDidRecordClick:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"%s called", __PRETTY_FUNCTION__);
}

- (void)adWillPresentFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"%s called", __PRETTY_FUNCTION__);
}

- (void)adWillDismissFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"%s called", __PRETTY_FUNCTION__);
}

- (void)adDidDismissFullScreenContent:(id<GADFullScreenPresentingAd>)ad {
  NSLog(@"%s called", __PRETTY_FUNCTION__);
  // Clear the rewarded interstitial ad.
  self.rewardedInterstitialAd = nil;
}

- (void)ad:(id)ad didFailToPresentFullScreenContentWithError:(NSError *)error {
  NSLog(@"%s called with error: %@", __PRETTY_FUNCTION__, error.localizedDescription);
}
// [END ad_events]

// [START show_ad]
- (void)showRewardedInterstitialAd {
  [self.rewardedInterstitialAd presentFromRootViewController:self
                                    userDidEarnRewardHandler:^{
                                      GADAdReward *reward = self.rewardedInterstitialAd.adReward;

                                      NSString *rewardMessage = [NSString
                                          stringWithFormat:@"Reward received with "
                                                           @"currency %@ , amount %ld",
                                                           reward.type, [reward.amount longValue]];
                                      NSLog(@"%@", rewardMessage);
                                      // TODO: Reward the user.
                                    }];
}
// [END show_ad]

@end

File: Swift/advanced/APIDemo/APIDemo/Snippets/RewardedInterstitialAdSnippets.swift
//
//  Copyright (C) 2025 Google, Inc.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.
//

import GoogleMobileAds

private class RewardedInterstitialAdSnippets: UIViewController, FullScreenContentDelegate {

  private var rewardedInterstitialAd: RewardedInterstitialAd?

  // Replace this ad unit ID with your own ad unit ID.
  private let adUnitID = "ca-app-pub-3940256099942544/6978759866"
  private let adManagerAdUnitID = "/21775744923/example/rewarded-interstitial"

  // [START load_ad]
  func loadRewardedInterstitialAd() async {
    do {
      rewardedInterstitialAd = try await RewardedInterstitialAd.load(
        // Replace this ad unit ID with your own ad unit ID.
        with: adUnitID, request: Request())
      // [START set_delegate]
      rewardedInterstitialAd?.fullScreenContentDelegate = self
      // [END set_delegate]
    } catch {
      print("Rewarded ad failed to load with error: \(error.localizedDescription)")
    }
  }
  // [END load_ad]

  // [START load_ad_gam]
  func loadAdManagerRewardedInterstitialAd() async {
    do {
      rewardedInterstitialAd = try await RewardedInterstitialAd.load(
        // Replace this ad unit ID with your own ad unit ID.
        with: adUnitID, request: AdManagerRequest())
      rewardedInterstitialAd?.fullScreenContentDelegate = self
    } catch {
      print("Rewarded ad failed to load with error: \(error.localizedDescription)")
    }
  }
  // [END load_ad_gam]

  // [START validate_server_side_verification]
  private func validateServerSideVerification() async {
    do {
      rewardedInterstitialAd = try await RewardedInterstitialAd.load(
        // Replace this ad unit ID with your own ad unit ID.
        with: adUnitID, request: Request())
      let options = ServerSideVerificationOptions()
      options.customRewardText = "SAMPLE_CUSTOM_DATA_STRING"
      rewardedInterstitialAd?.serverSideVerificationOptions = options
    } catch {
      print("Rewarded ad failed to load with error: \(error.localizedDescription)")
    }
  }
  // [END validate_server_side_verification]

  // [START validate_gam_server_side_verification]
  private func validateAdManagerServerSideVerification() async {
    do {
      rewardedInterstitialAd = try await RewardedInterstitialAd.load(
        // Replace this ad unit ID with your own ad unit ID.
        with: adUnitID, request: AdManagerRequest())
      let options = ServerSideVerificationOptions()
      options.customRewardText = "SAMPLE_CUSTOM_DATA_STRING"
      rewardedInterstitialAd?.serverSideVerificationOptions = options
    } catch {
      print("Rewarded ad failed to load with error: \(error.localizedDescription)")
    }
  }
  // [END validate_gam_server_side_verification]

  // [START ad_events]
  func adDidRecordImpression(_ ad: FullScreenPresentingAd) {
    print("\(#function) called.")
  }

  func adDidRecordClick(_ ad: FullScreenPresentingAd) {
    print("\(#function) called.")
  }

  func adWillPresentFullScreenContent(_ ad: FullScreenPresentingAd) {
    print("\(#function) called.")
  }

  func adWillDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
    print("\(#function) called.")
  }

  func adDidDismissFullScreenContent(_ ad: FullScreenPresentingAd) {
    print("\(#function) called.")
    // Clear the rewarded interstitial ad.
    rewardedInterstitialAd = nil
  }

  func ad(
    _ ad: FullScreenPresentingAd,
    didFailToPresentFullScreenContentWithError error: Error
  ) {
    print("\(#function) called with error: \(error.localizedDescription).")
  }
  // [END ad_events]

  // [START show_ad]
  func showRewardedInterstitialAd() {
    guard let rewardedInterstitialAd = rewardedInterstitialAd else {
      return print("Ad wasn't ready.")
    }

    // The UIViewController parameter is an optional.
    rewardedInterstitialAd.present(from: nil) {
      let reward = rewardedInterstitialAd.adReward
      print("Reward received with currency \(reward.amount), amount \(reward.amount.doubleValue)")
      // TODO: Reward the user.
    }
  }
  // [END show_ad]
}
